File: src/main/java/net/gtaun/shoebill/streamer/data/DynamicArea.java
package net.gtaun.shoebill.streamer.data;

import net.gtaun.shoebill.data.Vector3D;
import net.gtaun.shoebill.object.Destroyable;
import net.gtaun.shoebill.object.Player;
import net.gtaun.shoebill.object.Vehicle;
import net.gtaun.shoebill.streamer.Functions;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

/**
 * Created by Valeriy on 18/4/2016.
 */
public abstract class DynamicArea implements Destroyable {
    protected static Collection<DynamicArea> areas;

    static {
        areas = new ArrayList<>();
    }

    public static Set<DynamicArea> get() {
        return new HashSet<>(areas);
    }

    public static DynamicArea get(int id) {
        for (DynamicArea area : areas) {
            if(area.getId() == id) {
                return area;
            }
        }
        return null;
    }

    private int id, playerId;

    DynamicArea(int id, int playerId) {
        this.id = id;
        this.playerId = playerId;
    }

    public boolean isPlayerInRange(Player player) {
        return Functions.isPlayerInDynamicArea(player.getId(), this);
    }

    public boolean isPointInRange(Vector3D point) {
        return Functions.isPointInDynamicArea(this, point);
    }

    public void attachToPlayer(Player player, Vector3D offset) {
        Functions.attachDynamicAreaToPlayer(this, player, offset);
    }

    public void attachToObject(DynamicObject object, Vector3D offset) {
        Functions.attachDynamicAreaToObject(this, object, offset);
    }

    public void attachToVehicle(Vehicle vehicle, Vector3D offset) {
        Functions.attachDynamicAreaToVehicle(this, vehicle, offset);
    }

    public int getId() {
        return id;
    }

    public Player getPlayer() {
        return Player.get(playerId);
    }

    @Override
    public void destroy() {
        if(isDestroyed()) {
            selfRemove();
            return;
        }

        Functions.destroyDynamicArea(this);
        this.id = -1;
        selfRemove();
    }

    private void selfRemove() {
        if(areas.contains(this))
            areas.remove(this);
    }

    @Override
    public boolean isDestroyed() {
        return !Functions.isValidDynamicArea(this);
    }
}

File: src/main/java/net/gtaun/shoebill/streamer/data/DynamicArea.kt
package net.gtaun.shoebill.streamer.data

import net.gtaun.shoebill.entities.Destroyable
import net.gtaun.shoebill.entities.Player
import net.gtaun.shoebill.entities.Vehicle
import net.gtaun.shoebill.data.Vector3D
import net.gtaun.shoebill.streamer.AllOpen
import net.gtaun.shoebill.streamer.Functions
import net.gtaun.shoebill.streamer.Streamer
import net.gtaun.shoebill.streamer.event.area.PlayerEnterDynamicAreaEvent
import net.gtaun.shoebill.streamer.event.area.PlayerLeaveDynamicAreaEvent
import net.gtaun.util.event.Attentions
import net.gtaun.util.event.EventHandler
import net.gtaun.util.event.HandlerEntry
import net.gtaun.util.event.HandlerPriority
import java.util.*

/**
 * Created by Valeriy on 18/4/2016 in project streamer-wrapper.
 * Copyright (c) 2016 Valeriy. All rights reserved.
 */
@AllOpen
abstract class DynamicArea internal constructor(id: Int, val player: Player?) : Destroyable {

    final var id: Int = id
        get
        private set

    fun isPlayerInRange(player: Player): Boolean =
            Functions.isPlayerInDynamicArea(player.id, this)

    fun isPointInRange(point: Vector3D): Boolean =
            Functions.isPointInDynamicArea(this, point)

    fun attachToPlayer(player: Player, offset: Vector3D) =
            Functions.attachDynamicAreaToPlayer(this, player, offset)

    @JvmOverloads
    fun attachToObject(obj: DynamicObject, objectType: StreamerObjectType = StreamerObjectType.DYNAMIC,
                       forPlayer: Player ? = null, offset: Vector3D) =
            Functions.attachDynamicAreaToObject(this, obj, objectType, forPlayer?.id ?: 0xFFFF, offset)

    fun attachToVehicle(vehicle: Vehicle, offset: Vector3D) =
            Functions.attachDynamicAreaToVehicle(this, vehicle, offset)

    override fun destroy() {
        if (isDestroyed)
            return

        Functions.destroyDynamicArea(this)
        id = -1
        eventHandlers.forEach { it.cancel() }
        eventHandlers.clear()
        selfRemove()
    }

    private var eventHandlers: MutableList<HandlerEntry> = mutableListOf()

    fun playerEnterArea(handler: EventHandler<PlayerEnterDynamicAreaEvent>): HandlerEntry {
        val entry = eventManagerNode.registerHandler(PlayerEnterDynamicAreaEvent::class.java, handler, HandlerPriority.NORMAL,
                Attentions.create().`object`(this))
        eventHandlers.add(entry)
        return entry
    }

    fun playerEnterArea(handler: (PlayerEnterDynamicAreaEvent) -> Unit): HandlerEntry =
            playerEnterArea(EventHandler { handler(it) })

    fun playerLeaveArea(handler: EventHandler<PlayerLeaveDynamicAreaEvent>): HandlerEntry {
        val entry = eventManagerNode.registerHandler(PlayerLeaveDynamicAreaEvent::class.java, handler, HandlerPriority.NORMAL,
                Attentions.create().`object`(this))
        eventHandlers.add(entry)
        return entry
    }

    fun playerLeaveArea(handler: (PlayerLeaveDynamicAreaEvent) -> Unit): HandlerEntry =
            playerLeaveArea(EventHandler { handler(it) })

    private fun selfRemove() = areas.remove(this)

    override val isDestroyed: Boolean
        get() = !Functions.isValidDynamicArea(this)

    companion object {
        @JvmStatic
        protected var areas = mutableListOf<DynamicArea>()

        @JvmStatic
        fun get(): Set<DynamicArea> = HashSet(areas)

        private val eventManagerNode = Streamer.get().eventManager

        @JvmStatic
        operator fun get(id: Int): DynamicArea? = areas.find { it.id == id }
    }
}
